src/main/kotlin/com/noumenadigital/npl/cli/commands/registry/PumlCommand.kt

```
@@ -0,0 +1,82 @@
+import com.noumenadigital.npl.cli.service.CompilerService
+import com.noumenadigital.npl.cli.service.SourcesManager
+import com.noumenadigital.pumlgen.NplPumlObjectGenerator.generateFiles
+import java.io.File
+
+data class PumlCommand(
+    private val srcDir: String = ".",
+    private val outputDir: String = ".",
+) : CommandExecutor {
+    override val commandName: String = "puml"
+    override val description: String = "Generate a puml diagram from source in the given directory"
+
+    override val parameters: List<CommandParameter> =
+        listOf(
+            PositionalParameter(
+                name = "directory",
+                description = "Source directory containing NPL protocols",
+                defaultValue = ".",
+                isRequired = false,
+            ),
+        )
+
+    override fun execute(output: ColorWriter): ExitCode {
+        try {
+            if (!File(srcDir).isDirectory() || !File(srcDir).exists()) {
+                throw PumlError(GENERAL_ERROR, "Source directory does not exist or is not a directory: $srcDir")
+            }
+
+            val outputDir = File(outputDir).resolve("puml")
+            val protosMap = CompilerService(SourcesManager(srcDir)).compileAndReport(output = output).userDefinedMap
+
+            if (protosMap == null) {
+                throw PumlError(COMPILATION_ERROR, "No user defined types found, check sources and try again.")
+            }
+
+            val pumlFiles = generateFiles(protosMap)
+
+            if (!outputDir.mkdirs()) {
+                throw PumlError(INTERNAL_ERROR, "Failed to create output directory: ${outputDir.canonicalPath}")
+            }
+
+            output.info("Writing Puml files to ${outputDir.canonicalPath}\n")
+
+            pumlFiles.forEach {
+                try {
+                    it.write(outputDir.toPath())
+                } catch (e: Exception) {
+                    throw PumlError(INTERNAL_ERROR, e.message!!)
```

https://github.com/NoumenaDigital/npl-cli/blob/d84bbb2e2b2167f6e5f8d37a54bcb61da08cfab4/src/main/kotlin/com/noumenadigital/npl/cli/exception/InternalExceptions.kt#L13 you can use that one for example.

---

it's like here you can catch exception throw CommandExecutionException and in global error handling catch all exceptions and just rethrow them to the top level( there should be already all handlers impletented)

---

done. I should have looked at some of the other commands.